### DISCUSSION TITLE
DataGridColumnHeader not applied when default style present

### DISCUSSION THREAD

**Discussion by brinko99** [dotnet/wpf#7761] [Q&A]
I've been trying to figure out why a custom `DataGridColumnHeader` style will not be applied when a default header style is also present. Any help is appreciated.

Per the example below, when I apply the `DataGridTallStyle` to a `DataGrid`, the `DataGridColumnHeaderTallStyle` is not applied.

If `DataGridColumnHeaderTallStyle` is applied directly to the `DataGrid` it works as expected. The same situation with rows works in all cases.

See the attached project for a repo.

Thanks

```
<Style x:Key="{x:Type DataGridRow}" TargetType="{x:Type DataGridRow}">
    <Setter Property="MinHeight" Value="28" />
</Style>

<Style x:Key="DataGridRowTallStyle" TargetType="{x:Type DataGridRow}" BasedOn="{StaticResource {x:Type DataGridRow}}">
    <Setter Property="MinHeight" Value="55" />
</Style>

<Style x:Key="{x:Type DataGridColumnHeader}" TargetType="{x:Type DataGridColumnHeader}">
    <Setter Property="MinHeight" Value="28" />
</Style>

<Style x:Key="DataGridColumnHeaderTallStyle" TargetType="DataGridColumnHeader" BasedOn="{StaticResource {x:Type DataGridColumnHeader}}">
    <Setter Property="MinHeight" Value="55" />
</Style>

<Style x:Key="DataGridTallStyle" TargetType="{x:Type DataGrid}" BasedOn="{StaticResource {x:Type DataGrid}}">
    <Setter Property="ColumnHeaderStyle" Value="{StaticResource DataGridColumnHeaderTallStyle}"/>
    <Setter Property="RowStyle" Value="{StaticResource DataGridRowTallStyle}"/>
</Style>
```

```
<!--No DataGridColumnHeaderTallStyle applied-->
<DataGrid Style="{StaticResource DataGridTallStyle}"/>

<!--Works as expected-->
<DataGrid Style="{StaticResource DataGridTallStyle}"
          ColumnHeaderStyle="{StaticResource DataGridColumnHeaderTallStyle}"/>
```

[WpfDataGridColumnHeaderRepo.zip](https://github.com/dotnet/wpf/files/11326341/WpfDataGridColumnHeaderRepo.zip)

**Comment by miloush** [comment]
https://learn.microsoft.com/en-us/dotnet/desktop/wpf/properties/dependency-property-value-precedence states that implicit style has precedence over style setter value. Local value has precedence over implicit style. That seems to describe what is happening, or did I get something wrong?

**Reply by brinko99** [reply]
@miloush Thanks for the reply. Mapping that precedence list into real-world effects has always been a challenge for me.

In this specific instance, you'll see that an implicit style for `DataGridRow` is also being defined with a `MinHeight`. However, the `MinHeight` as applied by `DataGridRowTallStyle` is being used.

That is, using `<DataGrid Style="{StaticResource DataGridTallStyle}"/>` does indeed cause `DataGridRowTallStyle` to be applied, but not `DataGridColumnHeaderTallStyle`

Why might this work for the row but not the column header?

**Reply by miloush** [reply]
Right I see the confusion here. Row is a bit special in that it is the item of the items control (the `ItemsControl.ItemContainerStyle` is coerced to `RowStyle`). The item container generator therefore effectively applies the `RowStyle` explicitly in code:

https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/ItemsControl.cs,3450

which makes it a local value, regardless of how the row style was originally set. That does not happen to the header style

**Reply by brinko99** [reply]
Yes, I see the special case with the `RowStyle`.

I still find myself a bit confused over the precedence acting on the header. I'd assumed that the issue was the fact that the implicit header style defined the `MinHeight` value which took precedence over that in my named style.

However, even after removing the setter, the effect remains.
```
<Style x:Key="{x:Type DataGridColumnHeader}" TargetType="{x:Type DataGridColumnHeader}">
    <!--<Setter Property="MinHeight" Value="28" />-->
</Style>
```

Is it really that attempting to set the `ColumnHeaderStyle` of `DataGrid` itself which is losing out in the precedence hierarchy? If so, it seems that setting a style from within a style is a no-no if there is an implicit style for a given type. True?

`<Setter Property="ColumnHeaderStyle" Value="{StaticResource DataGridColumnHeaderTallStyle}"/>`

Thanks again
